File: fe/agents/shell/src/__tests__/statusDot.test.ts
import { describe, test, expect } from "vitest";
import { statusDotClass, effectiveStatus, applyEvent, type DotState } from "../statusDot.js";

/* These mirror internal/tools/agents/view/statusdot_test.go. The server
   paints the first frame and this code repaints it from live events, so a
   drift between them shows up as a dot that changes the moment an
   unrelated event arrives. */

describe("effectiveStatus", () => {
  const cases: [string, DotState, string][] = [
    ["leader working wins over child", { lifecycle: "working", subAgent: "working" }, "working"],
    ["leader spawning wins over child", { lifecycle: "spawning", subAgent: "working" }, "spawning"],
    ["idle leader with working child", { lifecycle: "idle", subAgent: "working" }, "subagent"],
    ["idle leader with spawning child", { lifecycle: "idle", subAgent: "spawning" }, "subagent"],
    ["killed leader with working child", { lifecycle: "killed", subAgent: "working" }, "subagent"],
    ["no leader process, working child", { lifecycle: "", subAgent: "working" }, "subagent"],
    ["idle leader, no child work", { lifecycle: "idle", subAgent: "" }, "idle"],
    ["killed leader alone stays blank", { lifecycle: "killed", subAgent: "" }, ""],
    ["nothing at all", { lifecycle: "", subAgent: "" }, ""],
  ];

  for (const [name, state, want] of cases) {
    test(name, () => {
      expect(effectiveStatus(state)).toBe(want);
    });
  }
});

describe("statusDotClass", () => {
  test("the sub-agent dot spins but is not the leader's green", () => {
    const sub = statusDotClass("subagent");
    expect(sub).toContain("animate-spin");
    expect(sub).not.toContain("border-green-500");
    expect(sub).not.toBe(statusDotClass("working"));
    // Same geometry as the other spinners, so rows do not jump on a change.
    expect(sub).toContain("h-3 w-3");
  });

  test("an unknown status stays hidden and still", () => {
    expect(statusDotClass("")).toContain("hidden");
    expect(statusDotClass("")).not.toContain("animate-spin");
  });
});

describe("applyEvent", () => {
  // The bug this guards: the sidebar used to hold one string per row, so a
  // leader's "idle" event overwrote everything — erasing a running
  // sub-agent's marker at the exact moment it became the only thing worth
  // showing.
  test("a leader transition leaves the sub-agent marker intact", () => {
    const prev: DotState = { lifecycle: "working", subAgent: "working" };
    const next = applyEvent(prev, { lifecycle: "idle" });
    expect(next).toEqual({ lifecycle: "idle", subAgent: "working" });
    expect(effectiveStatus(next)).toBe("subagent");
  });

  test("a sub-agent transition leaves the leader's lifecycle intact", () => {
    const prev: DotState = { lifecycle: "idle", subAgent: "" };
    const next = applyEvent(prev, { sub_agent: "working" });
    expect(next).toEqual({ lifecycle: "idle", subAgent: "working" });
  });

  test("a child that stopped clears only the sub-agent half", () => {
    const prev: DotState = { lifecycle: "idle", subAgent: "working" };
    const next = applyEvent(prev, { sub_agent: "" });
    expect(next).toEqual({ lifecycle: "idle", subAgent: "" });
    expect(effectiveStatus(next)).toBe("idle");
  });

  test("an event with neither key clears the sub-agent half", () => {
    // The server omits sub_agent entirely when it is empty (omitempty), so
    // a child that stopped arrives as a bare {type:"lifecycle"}.
    const prev: DotState = { lifecycle: "idle", subAgent: "working" };
    expect(applyEvent(prev, {})).toEqual({ lifecycle: "idle", subAgent: "" });
  });

  test("builds from nothing on a first event", () => {
    expect(applyEvent(undefined, { lifecycle: "working" })).toEqual({
      lifecycle: "working",
      subAgent: "",
    });
  });
});

File: internal/tools/agents/view/statusdot_test.go
package view

import (
	"strings"
	"testing"
)

// A conversation whose own process has gone idle but whose sub-agent is
// still grinding away used to render an idle dot — indistinguishable from
// a conversation with nothing running at all. The sidebar is the only
// place liveness is visible once you navigate away from a chat, so a
// child's work has to reach its parent's row.
func TestEffectiveStatusRollsUpSubAgentWork(t *testing.T) {
	cases := []struct {
		name      string
		lifecycle string
		status    string
		sub       string
		want      string
	}{
		// The leader owns the row whenever it is doing anything itself:
		// "my process is working" is a stronger statement than "something
		// downstream is", and conflating them would hide the difference.
		{"leader working wins over child", "working", "", "working", "working"},
		{"leader spawning wins over child", "spawning", "", "working", "spawning"},

		// The case this test exists for.
		{"idle leader with working child", "idle", "", "working", "subagent"},
		{"idle leader with spawning child", "idle", "", "spawning", "subagent"},
		{"killed leader with working child", "killed", "", "working", "subagent"},
		{"no leader process, working child", "", "", "working", "subagent"},

		// A queued child has nothing running yet. Reporting it as busy
		// would promise progress that is not happening — the same reason
		// isSubAgentWorking refuses to spin on a queued row.
		{"queued child is not work", "idle", "", "queued", "idle"},
		{"no child work leaves leader alone", "idle", "", "", "idle"},
		{"queued leader still shows", "", "queued", "", "queued"},
		{"nothing at all", "", "", "", ""},
		{"killed leader alone stays blank", "killed", "", "", ""},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			if got := effectiveStatus(tc.lifecycle, tc.status, tc.sub); got != tc.want {
				t.Errorf("effectiveStatus(%q, %q, %q) = %q, want %q",
					tc.lifecycle, tc.status, tc.sub, got, tc.want)
			}
		})
	}
}

// The dot has to be visibly different for delegated work, not just
// present: a green spinner on the parent row would say "this chat is
// working" when the chat itself is idle.
func TestStatusDotClassDistinguishesSubAgent(t *testing.T) {
	leader := StatusDotClass("working")
	sub := StatusDotClass("subagent")

	if !strings.Contains(sub, "animate-spin") {
		t.Errorf("sub-agent dot must spin, got %q", sub)
	}
	if sub == leader {
		t.Error("sub-agent dot is identical to the leader's — the row cannot say whose work it is")
	}
	if strings.Contains(sub, "border-green-500") {
		t.Errorf("sub-agent dot reuses the leader's green, got %q", sub)
	}
	// Still a dot on the same 12px geometry as the other spinners, so the
	// rows do not jump when a status changes.
	if !strings.Contains(sub, "h-3 w-3") {
		t.Errorf("sub-agent dot changed size, got %q", sub)
	}
	if strings.Contains(StatusDotClass(""), "animate-spin") {
		t.Error("empty status must not spin")
	}
	if !strings.Contains(StatusDotClass(""), "hidden") {
		t.Error("empty status must stay hidden")
	}
}
